src/main/scala/shine/cuda/primitives/intermediate/MapI.scala

```diff
@@ -0,0 +1,28 @@
+package shine.cuda.primitives.intermediate
+
+import shine.DPIA.DSL._
+import shine.DPIA.Phrases.Phrase
+import shine.DPIA.Types.DataType.idx
+import shine.DPIA.Types._
+import shine.DPIA._
+import shine.OpenCL._
+import shine.cuda.primitives.imperative.{ParFor, SyncThreads, SyncWarp}
+
+final case class MapI(level: ParallelismLevel, dim: Char) {
+  def apply(n: Nat, dt1: DataType, dt2: DataType,
+            f: Phrase[ExpType ->: AccType ->: CommType],
+            in: Phrase[ExpType],
+            out: Phrase[AccType]): Phrase[CommType] = {
+    val imperativ = comment(s"map${level.toString}") `;`
+                      ParFor(level, dim, unroll = false)(n, dt2, out,
+                        λ(expT(idx(n), read))(i => λ(accT(dt2))(a => f(in `@` i)(a))))
+    //TODO use other InsertMemoryBarrieres-mechanism
```

Comment:
are there difficulties to do this?

Reply:
A few PRs ago someone improved the way how to insert memory barriers in OpenCL. This should work for CUDA too, but i did not start to do this yet.

Reply:
mentioned in #106